src/nxt_http_return.c

```diff
@@ -87,27 +88,139 @@ nxt_http_return(nxt_task_t *task, nxt_http_request_t *r,
         return NULL;
     }
 
+    ctx = nxt_mp_zget(r->mem_pool, sizeof(*ctx));
+    if (nxt_slow_path(ctx == NULL)) {
+        nxt_http_request_error(task, r, NXT_HTTP_INTERNAL_SERVER_ERROR);
+        return NULL;
+    }
+
+    ctx->action = action;
     r->status = conf->status;
     r->resp.content_length_n = 0;
 
-    if (conf->location.length > 0) {
-        field = nxt_list_zero_add(r->resp.fields);
-        if (nxt_slow_path(field == NULL)) {
+    if (conf->loc_is_const) {
+        enc = nxt_str_dup(r->mem_pool, &ctx->encoded, &conf->encoded);
+        if (nxt_slow_path(enc == NULL)) {
             nxt_http_request_error(task, r, NXT_HTTP_INTERNAL_SERVER_ERROR);
             return NULL;
         }
 
-        nxt_http_field_name_set(field, "Location");
+        nxt_http_return_send_ready(task, r, ctx);
+
+    } else {
+        ret = nxt_var_query_init(&r->var_query, r, r->mem_pool);
+        if (nxt_slow_path(ret != NXT_OK)) {
+            nxt_http_request_error(task, r, NXT_HTTP_INTERNAL_SERVER_ERROR);
+            return NULL;
+        }
 
-        field->value = conf->location.start;
-        field->value_length = conf->location.length;
+        nxt_var_query(task, r->var_query, conf->location, &ctx->location);
+
+        nxt_var_query_resolve(task, r->var_query, ctx,
+                              nxt_http_return_send_ready,
+                              nxt_http_return_var_error);
+    }
+
+    return NULL;
+}
+
+
+static nxt_int_t
+nxt_http_return_encode(nxt_mp_t *mp, nxt_str_t *encoded,
+    const nxt_str_t *location)
+{
+    nxt_str_t   *loc;
+    nxt_uint_t  encode;
+
+    if (nxt_is_complex_uri_encoded(location->start, location->length)) {
+        loc = nxt_str_dup(mp, encoded, location);
+        if (nxt_slow_path(loc == NULL)) {
+            return NXT_ERROR;
+        }
+
+        return NXT_OK;
+    }
+
+    encode = nxt_encode_complex_uri(NULL, location->start, location->length);
+    encoded->length = location->length + encode * 2;
+
+    encoded->start = nxt_mp_nget(mp, encoded->length);
+    if (nxt_slow_path(encoded->start == NULL)) {
+        return NXT_ERROR;
     }
 
+    nxt_encode_complex_uri(encoded->start, location->start, location->length);
+
+    return NXT_OK;
+}
+
+
+static void
+nxt_http_return_send_ready(nxt_task_t *task, void *obj, void *data)
+{
+    nxt_int_t               ret;
+    nxt_str_t               *loc;
+    nxt_http_field_t        *field;
+    nxt_http_action_t       *action;
+    nxt_http_request_t      *r;
+    nxt_http_return_ctx_t   *ctx;
+    nxt_http_return_conf_t  *conf;
+
+    r = obj;
+    ctx = data;
+    action = ctx->action;
+    conf = action->u.conf;
+
```

Comment:
It's just a suggestion.
```
if (conf->encoded.length > 0) {
     // add location header.
}

if (conf->location != NULL && ctx->location.length > 0) {
         nxt_http_return_encode(r->mem_pool, &encoded, &ctx->location);

         // add location header.
}
```

It seems `conf->loc_is_const` and `ctx->encoded` can be elimiated.

Reply:
Yeah, maybe with the other logic that you put above I can remove it.  Will try.

Reply:
Actually I just realized that I had done a redundant copy here.